packages/daimo-api/src/contract/ClogMatcher.ts

```diff
@@ -0,0 +1,309 @@
+import { assertNotNull, debugJson, retryBackoff } from "@daimo/common";
+import { zeroAddr } from "@daimo/contract";
+import { daimoFastCctpAddress } from "@daimo/userop/dist/daimoOpSender";
+import { Kysely } from "kysely";
+import { Address, bytesToHex, getAddress, Hex, hexToBytes } from "viem";
+
+import { ForeignTokenTransfer } from "./foreignCoinIndexer";
+import { DB as IndexDB } from "../codegen/dbIndex";
+import { chainConfig } from "../env";
+import { TokenRegistry } from "../server/tokenRegistry";
+
+/**
+ * Finds corresponding information for a given transfer over logs in a tx.
+ * For example, a SwapClog and a cross-chain transfer both need to retrieve
+ * the actual "to" transfer.
+ */
+export class ClogMatcher {
+  /** Map of transactionHash to transfers within that transaction */
+  private txHashToTransfers: Map<string, ForeignTokenTransfer[]> = new Map();
+
+  /** Map of starting "to" transfer to the corresponding "from" transfer */
+  private correspondingSwapTransfers: Map<
+    ForeignTokenTransfer,
+    ForeignTokenTransfer
+  > = new Map();
```

Comment:
both of those data structures will keep growing as long as the server's running, right?

Reply:
yes -- txToTransfers grows... maybe instead should fetch on demand instead of in-memory? wdyt